### ISSUE TITLE
Updating user entity creates new token

### ISSUE THREAD

**Issue by danielchodusov** [api-platform/api-platform#1218]
Hi,

I've found a strange behaviour when updating an user entity with the PUT request. I'm signed in as "admin@example.com" using standard Symfony auth mechanism and when editing myself, trying to set invalid e-mail address:

```
{"@context":"/api/contexts/User","@id":"/api/users/1","@type":"http://schema.org/Person","displayName":"Admin", "email":"invalid email"}
```

Symfony Validator correctly return the reponse "invalid e-mail" for the "email" property - but I'm also signed off because a new token is created "with invalid email" as the username.

<img width="1107" alt="Screenshot 2019-08-12 at 12 55 49" src="https://user-images.githubusercontent.com/4260378/62860470-904b1480-bd00-11e9-889b-cbdc95236349.png">

This behaviour happens before persisting changed e-mail to the database, when I send the valid e-mail address, it happens as well. A new token is created whenever I send serialized user with the updated e-mail adress.

Anyone have a clue why/how this might happen?

**Comment by danielchodusov** [comment]
I have resolved this, by implementing \EquatableInterface to the user. Thought it's seems to be strange, that the new token is created before persisting to entity, this have fixed my issue.

**Comment by dunglas** [comment]
I think that I understand what happens, and it looks very dangerous:

* the in memory Doctrine entity representing the user is updated, because it's how API Platform works
* because of the validation error, it isn't persisted in database (intended)
* however, the Symfony security component uses the in memory Doctrine entity as the current user for the token, it the **same** object that is used (both id), it means that at some point the token contains the (not validated) data

I'm not sure of all the implications yet, but it looks bad so reopening.

**Comment by danielchodusov** [comment]
I can confirm:

Whenever I made a request which deserialize the entity specified in the user provider and this entity is my current user, I recieve a new token based on deserialization data.

```
providers:
    user_provider:
      entity:
        class: App\Entity\User\User
        property: email
```

**Comment by lwillems** [comment]
Hello

Same thing seems happening with roles and access_control.
Got a custom voter used in access_control for PUT on /users :  checking hierarchical roles to avoid user to self-upgrade his account if he's not allowed to perform this operation.
however if i send ROLE_ADMIN in PUT request, its replace role in entity and in token user (same id as @dunglas said) before reaching voter so user is already ADMIN when checking permission in voter.
Seems dangerous also.

**Comment by dunglas** [comment]
Fixed by #2992. Thanks for the report, it has been very useful!
